src/main/java/org/wildfly/discovery/Discovery.java

```diff
@@ -100,6 +100,36 @@ public ServicesQueue discover(ServiceType serviceType, FilterSpec filterSpec) {
         return new BlockingQueueServicesQueue(queue, problems, provider.discover(serviceType, filterSpec, result));
     }
 
+    /**
+     * Perform a service discovery.  The returned services queue is populated as discovery answers become available.
+     * Answers may be cached within each provider.  The order of answers is not significant and can vary from call to
+     * call, especially with asynchronous discovery mechanisms.  The returned service queue may be closed to indicate
+     * no further interest in query answers, and for this purpose it implements {@link AutoCloseable} in order to
+     * facilitate simple usage in a {@code try}-with-resources block.
+     *
+     * This method allows a timeout to be specified.
+     *
+     * @param serviceType the abstract or concrete type of service to search for
+     * @param filterSpec the service filter specification
+     * @param timeout the timeout duration
+     * @param timeUnit the unit of time for the timeout
+     * @return the services queue
+     */
+    public ServicesQueue discover(ServiceType serviceType, FilterSpec filterSpec, long timeout, TimeUnit timeUnit) {
+        if (timeout <= 0 || timeUnit == null) {
```

Comment:
I think it is not good for a timeout of 0 to mean "forever" as it can cause problems when things get rounded.  Instead a value of 0 or a `null` for `timeUnit` should be rejected; see `Assert.check*` for additional validation helpers.